language/libra-vm/src/logger.rs

```diff
@@ -0,0 +1,126 @@
+    // `StateViewId`
+
+    // StateViewId::BlockExecution - typical transaction execution
+    block_id: Option<HashValue>,
+    // StateViewId::ChunkExecution - state sync
+    first_version: Option<Version>,
+    // StateViewId::TransactionValidation - validation
+    base_version: Option<Version>,
+
+    // transaction position in the list of transactions in the block,
+    // 0 if the transaction is not part of a block (i.e. validation).
+    txn_id: usize,
+}
+
+impl LogSchema {
+    pub fn new(view_id: StateViewId, txn_id: usize) -> Self {
+        match view_id {
+            StateViewId::BlockExecution { block_id } => Self {
+                name: LogEntry::Execution,
+                block_id: Some(block_id),
+                first_version: None,
+                base_version: None,
+                txn_id,
+            },
+            StateViewId::ChunkExecution { first_version } => Self {
+                name: LogEntry::Execution,
+                block_id: None,
+                first_version: Some(first_version),
+                base_version: None,
+                txn_id,
+            },
+            StateViewId::TransactionValidation { base_version } => Self {
+                name: LogEntry::Validation,
+                block_id: None,
+                first_version: None,
+                base_version: Some(base_version),
+                txn_id,
+            },
+            StateViewId::Miscellaneous => Self {
+                name: LogEntry::Miscellaneous,
+                block_id: None,
+                first_version: None,
+                base_version: None,
+                txn_id,
+            },
+        }
+    }
+}
+
+#[derive(Clone, Copy, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum LogEntry {
+    Execution,
+    Validation,
+    Miscellaneous, // usually testing
+}
+
+#[derive(Clone)]
+pub struct LibraLogger {
+    schema: LogSchema,
+}
+
+impl LibraLogger {
+    pub(crate) fn new(view_id: StateViewId, txn_id: usize) -> Self {
+        Self {
+            schema: LogSchema::new(view_id, txn_id),
+        }
+    }
+}
+
+impl Logger for LibraLogger {
+    // A critical log for the adapter, increment the `CRITICAL_ERRORS` monitor event so we can
+    // hook an alert when a critical occurs
+    fn crit(&self, message: &str) {
+        CRITICAL_ERRORS.inc();
+        error!(self.schema.clone(), "{}", message);
+    }
+
+    // REVIEW: consider hooking up a monitor event that we can use with a threshold
+    // over a period of time to raise an alert
+    fn error(&self, message: &str) {
+        error!(self.schema.clone(), "{}", message);
+    }
+
+    fn warn(&self, message: &str) {
+        warn!(self.schema.clone(), "{}", message);
+    }
+
+    fn info(&self, message: &str) {
+        info!(self.schema.clone(), "{}", message);
+    }
+
+    fn debug(&self, message: &str) {
+        debug!(self.schema.clone(), "{}", message);
+    }
+}
```

Comment:
@dariorussi: This is sort of an anti-pattern w.r.t. logging and will probably result in things being much more difficult to understand and debug.

* The logging macros expand to capture the filename, module, and line number of where they're invoked. Wrapping the macros in a function call completely removes all of this information from the instrumentation location as now all logs in the VM will appear as if they're coming from this file instead of where they were invoked.

* We recently removed the `crit` log level from the rest of the system, it would probably be best to keep things similar across the stack by removing `crit` and supporting `trace` levels.

* Is there a particular reason why you wanted to thread through a logger via passing it in as a parameter instead of relying on the globally initialized one supplying by the logging facade in `libra-logger`?

More over, given you've started to go this route, is there something lacking from the logging framework that we could add that would better support your needs?

Reply:
> @dariorussi: This is sort of an anti-pattern w.r.t. logging and will probably result in things being much more difficult to understand and debug.
> 
> * The logging macros expand to capture the filename, module, and line number of where they're invoked. Wrapping the macros in a function call completely removes all of this information from the instrumentation location as now all logs in the VM will appear as if they're coming from this file instead of where they were invoked.

I know and we have talked about this in the channel. We can think to go down a different path but it's not an easy one and later you will find an explanation. Talking to @gregnazario we concluded that, though this is not the ideal for the reason you just mentioned it may not be a deal breaker as the line is not used in a critical way. To me I understand that for certain generic errors (any of those where the string is not enough to identify the location) having the line number matters. In the VM case often the line number is insignificant because it's the line in the Move code that matters and that is not and cannot be reflected in the macro.
Having said so, suggestions welcome

> * We recently removed the `crit` log level from the rest of the system, it would probably be best to keep things similar across the stack by removing `crit` and supporting `trace` levels.

As far as we are concerned the difference between critical and error is in the incrementing a counter that we pick up in monitoring for an alert. Se that difference is vital to us and as you probably noticed both `crit` and `error` go to `error!`. I would like to keep it that way for the VM as it is significant. Again suggestion welcome.

> * Is there a particular reason why you wanted to thread through a logger via passing it in as a parameter instead of relying on the globally initialized one supplying by the logging facade in `libra-logger`?
> 

There are information that the Move VM is unaware of. Specifically everything Libra related is unknown to the Move VM. In that sense it gets hard to push knowledge of Libra into the VM and having a logger controlled from the outside is the only reasonable way.
Thoughts?

> More over, given you've started to go this route, is there something lacking from the logging framework that we could add that would better support your needs?

To be honest the way I have done it makes a lot of sense to me. So I don't know that there is much missing. It may be more for you to evaluate...

Reply:
cc @vgao1996